onnxruntime/core/providers/coreml/builders/impl/pool_op_builder.cc

```diff
@@ -0,0 +1,89 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/providers/shared/utils/utils.h"
+#include "core/providers/coreml/builders/helper.h"
+#include "core/providers/coreml/builders/model_builder.h"
+#include "core/providers/coreml/builders/op_builder_factory.h"
+
+#include "base_op_builder.h"
+
+namespace onnxruntime {
+namespace coreml {
+
+class PoolOpBuilder : public BaseOpBuilder {
+  // Add operator related
+ private:
+  Status AddToModelBuilderImpl(ModelBuilder& model_builder, const Node& node,
+                               const logging::Logger& logger) const override ORT_MUST_USE_RESULT;
+
+  // Operator support related
+ private:
+  bool IsOpSupportedImpl(const InitializedTensorSet& initializers, const Node& node,
+                         const logging::Logger& logger) const override;
+};
+
+// Add operator related
+
+Status PoolOpBuilder::AddToModelBuilderImpl(ModelBuilder& model_builder,
+                                            const Node& node,
+                                            const logging::Logger& /* logger */) const {
+  std::unique_ptr<COREML_SPEC::NeuralNetworkLayer> layer = CreateNNLayer(node);
+
+  auto* coreml_pool = layer->mutable_pooling();
+  const auto& op_type = node.OpType();
+
+  // We only support global pool now
+  coreml_pool->set_globalpooling(true);
+  coreml_pool->mutable_valid();
+
+  if (op_type == "GlobalAveragePool") {
+    coreml_pool->set_type(COREML_SPEC::PoolingLayerParams_PoolingType_AVERAGE);
+  } else if (op_type == "GlobalMaxPool") {
+    coreml_pool->set_type(COREML_SPEC::PoolingLayerParams_PoolingType_MAX);
+  }
+
```

Comment:
Should there be an 'else' branch here to catch unexpected usage?

Reply:
Add it to the IsOpSupportedImpl() so this will never happen here